### ISSUE TITLE
How to retrieve entities (& relations) within angular & ui-router state provider resolve()?

### ISSUE THREAD

**Issue by mobidev111** [js-data/js-data#240]
#### Setup
- js-data
- js-data-localforage
- js-data-angular
- angular ui router

#### See plunker
http://plnkr.co/edit/LzHSHkdF9b4yk0r0vVp5?p=preview

1) Click button "Create User" to create a user
2) Click button "Users" to lists all users

#### Problem A: retrieving User entity with relations fails

```
User.findAll(null, {
 with: ['post']
})
```

fails with 

```
"TypeError: Cannot read property 'name' of undefined"
```

#### Problem B: newly created users only show up after page reload

tested without relations:
```
User.findAll()
```

leads to the problem that a newly created user only shows up after page reload

Adding the { bypassCache: true } doesn't make any difference.

**Comment by jmdobry** [comment]
You've discovered a quirk of Angular's DI, which I've documented for you [here](http://www.js-data.io/v2.6/docs/js-data-angular#section-angular-s-di-is-lazy). 

tldr; Inject `Post` into your route resolves and it will work.

**Comment by mobidev111** [comment]
Nice. This solves Problem A - it now retrieves the User entity.

Problem B persists - when I create a User entity, it only shows up in the user list after reloading the page.

**Comment by jmdobry** [comment]
Here's a fork of your plunker that shows a way to make it work: http://plnkr.co/edit/SnNHSGBFCBz76VPmRtnW?p=preview

My only guess as to why `bypassCache: true` didn't work is ui-router is doing some kind of caching the ngRouter doesn't do.

**Comment by mobidev111** [comment]
Excellent. bypassCache: true works now.

**Comment by mobidev111** [comment]
This means that I need to read directly from the adapter (with ```bypassCache: true```) for all find / findAll queries? And thus essentially can't leverage the in-memory store because it is never read?

Just being curious: What is the underlying cause for not being able to directly read a newly written entity?

**Comment by jmdobry** [comment]
The newly created value _is_ in the store, and if you ask the store for it directly, you'll get it. Your issue is that after ui-router resolves a state, it caches it, so even when you go to the users route again, ui-router gives you the value resolved the first time you visited the route. This is different default behavior than what ng-router does. You have to explicitly tell ui-router to reload states, and thus get the new data that's in the in-memory store. Also, make sure `DS.defaults.useFilter` is set to `true`.

Add `ui-sref-opts="{reload: true}"` to your `<a ui-sref="...">...</a>` and it should work without needing to set `bypassCache` to `true`.

**Comment by shairez** [comment]
@jmdobry would you say it happens only because of the resolve caching? 
Because it happens to me even when I'm not using `resolve`, 
I'm calling `findAll` in the constructor of the controller of the newly created state, and it still fetch the data from cache.

Is the expected behavior for the `create` method to invalidate the cache for later `find` calls?

**Comment by jmdobry** [comment]
Outside of ui-router's resolve caching, the behavior of [`DS#find`](http://www.js-data.io/docs/dsfind) and [`DS#findAll`](http://www.js-data.io/docs/dsfindall) is governed by the options: [`bypassCache`](http://www.js-data.io/docs/dsdefaults#bypasscache), [`cacheResponse`](http://www.js-data.io/docs/dsdefaults#cacheresponse), [`findStrictCache`](http://www.js-data.io/docs/dsdefaults#findstrictcache), [`useFilter`](http://www.js-data.io/docs/dsdefaults#usefilter), etc.

**Comment by shairez** [comment]
@jmdobry I understand.
My question is this (with the following scenario in mind) - 
1. I queried a list of members using `findAll`
2. I created a new member
3. Again, I queried using `findAll` for a list of members

Shouldn't I be getting the updated list of members without using any of the options you've listed?
Shouldn't it be the default behavior? 
What are the pros of not making it the default choice?

Thanks Jason!

**Comment by jmdobry** [comment]
In this scenario you're looking for the `useFilter` option, which defaults to `false`. Set it to `true` and you will get the behavior you desire. I personally thought the `userFilter` default should be `true`, but users clamored for `false`.

**Comment by shairez** [comment]
Thanks for your quick answer, I'm very curious to read what is the reasoning for defaulting to `false` but couldn't find it, do you have a link to those discussions? (or just where to search them? and I'll try my luck again)
Thanks again for your efforts and support!

**Comment by jmdobry** [comment]
Closing question issue.
